File: opensource/kernel/src/main/java/com/sos/scheduler/engine/kernel/settings/CppSettings.java
package com.sos.scheduler.engine.kernel.settings;

import com.sos.scheduler.engine.kernel.cppproxy.SettingsC;
import com.sos.scheduler.engine.kernel.settings.database.DatabaseSettings;

public final class CppSettings {
    private final SettingsC cppProxy;

    public CppSettings(SettingsC cppProxy) {
        this.cppProxy = cppProxy;
    }

    public void setCppSettings(Settings s) {
        setDatabaseCppSettings(s.getDatabaseSettings());
    }

    private void setDatabaseCppSettings(DatabaseSettings s) {
        if (s.getHostwarePathOrNull() != null)
            cppProxy.set_db_name(s.getHostwarePathOrNull());
    }
}

File: opensource/kernel/src/main/scala/com/sos/scheduler/engine/kernel/settings/CppSettings.scala
package com.sos.scheduler.engine.kernel.settings

import com.sos.scheduler.engine.kernel.cppproxy.SettingsC
import com.sos.scheduler.engine.kernel.settings.CppSettingName._

/**
 * C++ settings (settings.cxx), for write-access only.
 * For reading, see [[com.sos.scheduler.engine.kernel.scheduler.SchedulerConfiguration]].
 */
final class CppSettings private(val valueMap: Map[CppSettingName, String]) {

  def apply(name: CppSettingName) = valueMap(name)

  def setSettingsInCpp(cppProxy: SettingsC): Unit = {
    for (e <- valueMap)
      cppProxy.set(e._1.getNumber, e._2)
  }
}

object CppSettings {
  lazy val Empty = apply(Nil)
  lazy val TestMap = Map(alwaysCreateDatabaseTables -> true.toString)

  def apply(values: Iterable[(CppSettingName, String)]): CppSettings =
    new CppSettings(values.toMap withDefault { o ⇒ throw new NoSuchElementException(s"Unknown CppSettingName $o") })
}
